### ISSUE TITLE
Roslyn should adjust set values for members with outputs that are stricter than inputs after processing postconditions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#40894]
**Version Used**: 
```
commit a0404639df1ab23612390052d0b5d400c864bc38
Merge: eb26598979 6aa3e57ac0
Author: Sam Harwell <sam.harwell@microsoft.com>
Date:   Mon Dec 30 20:03:18 2019 -0800

    Merge pull request #40643 from sharwell/driver-tasks

    Reduce Task allocations in the analyzer driver
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class D
{
  string? field;
  [NotNull] public ref string? Prop
  {
    get
    {
      field = "";
      return ref field; 
    }
  }
  
  void M()
  {    
    Prop.ToString();
    Out1(out Prop);
    Prop.ToString();
    Out2(out Prop);
    Prop.ToString();
  }
  
  public void Out1(out string? s) => s = null;
  public void Out2([MaybeNullWhen(false)] out string? s) => s = null;
}
```

**Expected Behavior**:
No warnings

**Actual Behavior**:
`CS8602: Dereference of a possibly null reference.` is reported only for the last derefernece of Prop.

**Notes**
`Prop` always returns not null according to its annotation. Since allowed inputs include nulls Roslyn adjusts the set value for assignments to this property including `out` arguments. However the adjustment happens before Roslyn learns from methods' post-conditions which effectively overwrite what was assigned to the variable based on the parameters' annotations.
Since this does not affect the fact that `Prop` might be assigned with null and still return not-null values according to its annotations this seems to be a mistake.

If you wwi
